Source/DafnyLanguageServer/Handlers/DafnyHoverHandler.cs

```diff
@@ -24,30 +30,145 @@ public class DafnyHoverHandler : HoverHandlerBase {
+
+                var assertionId = assertionCount == 1 ? "" : $" #{assertionIndex + 1}/{assertionCount}";
+                var assertionInfo = $" of {batchRef} #{assertionBatchIndex + 1}/{assertionBatchCount} checked in {assertionBatchTime:n0}ms with {assertionBatchResourceCount:n0} resource count";
+
+                information += $"assertion{assertionId}{assertionInfo}:  \n";
+                // Not the main error displayed in diagnostics
+                if (!(assertionNode.GetCounterExample() is ReturnCounterexample returnCounterexample2 &&
+                      returnCounterexample2.FailingReturn.tok.GetLspRange().Contains(position))) {
+                  information += assertionNode.SecondaryPosition != null
+                    ? $"`{Path.GetFileName(assertionNode.Filename)}({assertionNode.SecondaryPosition.Line + 1}, {assertionNode.SecondaryPosition.Character + 1}): `"
+                    : "";
+                }
+
+                information += "*";
+                var statusMessage = assertionNode.StatusVerification switch {
+                  VerificationStatus.Verified => "Verified",
+                  VerificationStatus.Inconclusive => "Could not be checked",
+                  VerificationStatus.Error => assertCmd?.ErrorMessage ?? "Might not hold",
+                  _ => "Not checked",
+                };
+                var counterexample = assertionNode.GetCounterExample();
+
+                void SetDescription(ProofObligationDescription? description) {
+                  if (description == null) {
+                    return;
+                  }
+
+                  statusMessage = assertionNode?.StatusVerification switch {
+                    VerificationStatus.Verified => description.SuccessDescription,
+                    VerificationStatus.Error => description.FailureDescription,
+                    _ => statusMessage
+                  };
+                }
+
+                if (counterexample is ReturnCounterexample returnCounterexample) {
+                  SetDescription(returnCounterexample.FailingReturn.Description);
+                } else if (counterexample is CallCounterexample callCounterexample) {
+                  SetDescription(callCounterexample.FailingCall.Description);
+                }
+
+                information += statusMessage;
+                information += assertionNode.StatusCurrent switch {
+                  CurrentStatus.Current => "",
+                  CurrentStatus.Obsolete => "recently",
+                  _ => "recently and verifying..."
+                };
+                information += "*  \n";
+              }
+
+              assertionIndex++;
+            }
+          }
+
+          if (information != "") {
+            return information;
+          }
+          // Ok no assertion here. Maybe a method?
+          if (node.Position.Line == position.Line &&
+              node.Filename == document.Uri.GetFileSystemPath()) {
```

Comment:
For the line containing the method, hovering over the first line of a method provides statistics of this method